packages/react/src/useEditorState.ts

```diff
@@ -0,0 +1,125 @@
+  editor: TEditor;
+  transactionNumber: number;
+};
+export type UseEditorStateOptions<
+  TSelectorResult,
+  TEditor extends Editor | null = Editor | null,
+> = {
+  /**
+   * The editor instance.
+   */
+  editor: TEditor;
+  /**
+   * A selector function to determine the value to compare for re-rendering.
+   */
+  selector: (context: EditorStateSnapshot<TEditor>) => TSelectorResult;
+  /**
+   * A custom equality function to determine if the editor should re-render.
+   * @default `(a, b) => a === b`
+   */
+  equalityFn?: (a: TSelectorResult, b: TSelectorResult | null) => boolean;
+};
+
+/**
+ * To synchronize the editor instance with the component state,
+ * we need to create a separate instance that is not affected by the component re-renders.
+ */
+function makeEditorStateInstance<TEditor extends Editor | null = Editor | null>(initialEditor: TEditor) {
+  let transactionNumber = 0
+  let lastTransactionNumber = 0
+  let lastSnapshot: EditorStateSnapshot<TEditor> = { editor: initialEditor, transactionNumber: 0 }
+  let editor = initialEditor
+  const subscribers = new Set<() => void>()
+
+  const editorInstance = {
+    /**
+     * Get the current editor instance.
+     */
+    getSnapshot(): EditorStateSnapshot<TEditor> {
+      if (transactionNumber === lastTransactionNumber) {
+        return lastSnapshot
+      }
+      lastTransactionNumber = transactionNumber
+      lastSnapshot = { editor, transactionNumber }
+      return lastSnapshot
+    },
+    /**
+     * Always disable the editor on the server-side.
+     */
+    getServerSnapshot(): EditorStateSnapshot<null> {
+      return { editor: null, transactionNumber: 0 }
+    },
+    /**
+     * Subscribe to the editor instance's changes.
+     */
+    subscribe(callback: () => void) {
+      subscribers.add(callback)
+      return () => {
+        subscribers.delete(callback)
+      }
+    },
+    /**
+     * Watch the editor instance for changes.
+     */
+    watch(nextEditor: Editor | null) {
+      editor = nextEditor as TEditor
+
+      if (editor) {
+        /**
+         * This will force a re-render when the editor state changes.
+         * This is to support things like `editor.can().toggleBold()` in components that `useEditor`.
+         * This could be more efficient, but it's a good trade-off for now.
+         */
+        const fn = () => {
+          transactionNumber += 1
+          subscribers.forEach(callback => callback())
+        }
+
+        const currentEditor = editor
+
+        currentEditor.on('transaction', fn)
+        return () => {
+          currentEditor.off('transaction', fn)
+        }
+      }
+    },
+  }
+
+  return editorInstance
+}
+
+export function useEditorState<TSelectorResult>(
+  options: UseEditorStateOptions<TSelectorResult, Editor | null>
+): TSelectorResult | null;
+export function useEditorState<TSelectorResult>(
+  options: UseEditorStateOptions<TSelectorResult, Editor>
+): TSelectorResult;
```

Comment:
Do you mean `useEditor` or `useEditorState`?

Because if I go into node_modules and just flip the order of these overloads (in the `.d.ts` file of `useEditorState`) it seems to be working all fine. I think this is because TypeScript will resolve the `Editor` overload first, and if it's null it will resolve the `Editor | null` second. But since `Editor` DOES extend `Editor | null`, it resolved the nullable one first. Sorry that's a bit of a mouthful to read😆

I'm wondering if just flipping the order of the overloads in the source code will fix it? Would it be possible to make another changeset & pre-release to test?

Happy to make a PR as well if I haven't explained it very well 😄

Reply:
Oh I was talking useEditor, Hm interesting. I would happily take a PR for it!

Reply:
Gotcha. will make a PR

Fwiw, typescript IS able to infer the boolean to `true` exactly, but only because it was explicitly written out in the overloads. [Here's an example of that behaviour ](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgKIBNhgPZWQbwF8BuAKFIHoLkBBEZCADzgFsAHAGxWxmRG2TYAbtA7Y46AM7I4YGRw4AaPtjlxkAcygRZyACIANUuggIOcbchgBXEAjDBs9a5IgYsuAOpYAFgDlsAHkRKDEJAApsNgcnSQB+AC4CYBYWCExZCA4ATwAlCBATKETkACNsbC44EEIASiT3HDwAHz5rBTJKam8wH0EQsKllOGkTFliwKEz0ZFBkB2i4NmNTc0sbOxjnV0avX2DRcXRI6McQSST8FLSMyBz8wugkyesIOobMJrITMwsUDfsZ2QLjcnz2vQOoSOJy28Uu13SwEy9wKRRK8A4rneaDBLTaHXIVDQzHYXEkXSolKp1JptLp9IZjKZ1PICAmyAA7r4ApDBjRkABeYE7XE9fxBAbQ2pkZCyokAPTirPZXIhkok-KFIN2UDFvKlMrl1EVyvOclVPn1EgAQoLhaCPLr9urjldUojkXlUU8rHBMW9paRZcgFUrTZJzc7DhIAMJ27WiqNQiJum5Iu5ex5QZ5QV51Q0h41K4NEgAqPmA0icKDZUG09hysxAMGg0hwZRQACsXHIdUA)